pyface/ui/qt4/data_view/data_view_item_model.py

```diff
@@ -0,0 +1,235 @@
+            bottom_right = self._to_model_index(bottom, right)
+            self.dataChanged.emit(top_left, bottom_right)
+
+    # Structure methods
+
+    def parent(self, index):
+        if not index.isValid():
+            return QModelIndex()
+        parent = index.internalPointer()
+        if parent == Root:
+            return QModelIndex()
+
+        grandparent, row = self.model.index_manager.get_parent_and_row(parent)
+        return self.createIndex(row, 0, grandparent)
+
+    def index(self, row, column, parent):
+        if parent.isValid():
+            parent_index = self.model.index_manager.create_index(
+                parent.internalPointer(),
+                parent.row(),
+            )
+        else:
+            parent_index = Root
+        index = self.createIndex(row, column, parent_index)
+        return index
+
+    def rowCount(self, index):
+        row_index = self._to_row_index(index)
+        try:
+            if self.model.can_have_children(row_index):
+                return self.model.get_row_count(row_index)
+        except Exception:
+            logger.exception("Error in rowCount")
+
+        return 0
+
+    def columnCount(self, index):
+        row_index = self._to_row_index(index)
+        try:
+            return self.model.get_column_count(row_index) + 1
+        except Exception:
+            logger.exception("Error in columnCount")
+
+    # Data methods
+
+    def flags(self, index):
+        row = self._to_row_index(index)
+        column = self._to_column_index(index)
+        value_type = self.model.get_value_type(row, column)
+
+        flags = Qt.ItemIsEnabled | Qt.ItemIsSelectable
+        if is_qt5 and not self.model.can_have_children(row):
+            flags |= Qt.ItemNeverHasChildren
+
+        if value_type and value_type.can_edit(self.model, row, column):
+            flags |= Qt.ItemIsEditable
+
+        return flags
+
+    def data(self, index, role=Qt.DisplayRole):
+        row = self._to_row_index(index)
+        column = self._to_column_index(index)
+        value_type = self.model.get_value_type(row, column)
+        if not value_type:
+            return None
+
+        if role == Qt.DisplayRole:
+            if value_type.has_text(self.model, row, column):
+                return value_type.get_text(self.model, row, column)
+        elif role == Qt.EditRole:
+            if value_type.can_edit(self.model, row, column):
+                return value_type.get_editable(self.model, row, column)
+
+        return None
+
+    def setData(self, index, value, role=Qt.EditRole):
+        row = self._to_row_index(index)
+        column = self._to_column_index(index)
+        value_type = self.model.get_value_type(row, column)
+        if not value_type:
+            return False
+
+        if role == Qt.EditRole:
+            if value_type.can_edit(self.model, row, column):
+                return value_type.set_editable(self.model, row, column, value)
+        elif role == Qt.TextRole:
```

Comment:
Good catch.  Somewhat surprising that this is working.

Reply:
The next question is... why didn't the tests detect this? 🤔

Reply:
The tests didn't detect it because this is UI scaffolding code that is likely to be replaced/re-written.  And looking more closely, it probably never gets hit because the role is always `Qt.EditRole` when called from existing Qt objects.